core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushLimitThroughUnion.java

```diff
@@ -91,4 +94,24 @@ public Result apply(LimitNode parent, Captures captures, Context context)
                 parent.replaceChildren(ImmutableList.of(
                         unionNode.replaceChildren(builder.build()))));
     }
+
+    // Avoid inserting Limit nodes between nested Union nodes until after MergeUnion has run, otherwise the inserted
+    // LimitNode will prevent the merging from happening
+    private boolean isNestedUnionSourceNode(PlanNode node, Lookup lookup)
+    {
+        while (true) {
+            if (node instanceof UnionNode) {
+                return true;
+            }
+            else if (node instanceof ProjectNode project) {
```

Comment:
would this work if we only look for directly `UnionNode <- UnionNode` or projection match is needed?

Reply:
It does not. The `IterativeOptimizer` that this will now run in has lots of other rules to handle identity projections, but those aren't guaranteed to run in any particular order so we have to ignore those when looking for nested unions.